pkg/dockerfile/fast_generator.go

```diff
@@ -0,0 +1,273 @@
+	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
+		return "", fmt.Errorf("Failed to write %s: %w", filename, err)
+	}
+	if err := os.WriteFile(path, data, 0o644); err != nil {
+		return "", fmt.Errorf("Failed to write %s: %w", filename, err)
+	}
+	return path, nil
+}
+
+func (g *FastGenerator) generateMonobase(lines []string, tmpDir string) ([]string, error) {
+	lines = append(lines, []string{
+		"# syntax=docker/dockerfile:1-labs",
+		"FROM r8.im/monobase:latest",
+	}...)
+
+	cogPath, err := g.copyCog(tmpDir)
+	if err != nil {
+		return nil, err
+	}
+
+	lines = append(lines, []string{
+		"ENV R8_COG_VERSION=\"file:///buildtmp/" + filepath.Base(cogPath) + "\"",
+	}...)
+
+	skipCudaArg := "--skip-cuda"
+	if g.Config.Build.GPU {
+		skipCudaArg = ""
+		cudaVersion := g.Config.Build.CUDA
+		cudnnVersion := g.Config.Build.CuDNN
+		lines = append(lines, []string{
+			"ENV R8_CUDA_VERSION=" + cudaVersion,
+			"ENV R8_CUDNN_VERSION=" + cudnnVersion,
+			"ENV R8_CUDA_PREFIX=https://monobase.replicate.delivery/cuda",
+			"ENV R8_CUDNN_PREFIX=https://monobase.replicate.delivery/cudnn",
+		}...)
+	}
+
+	lines = append(lines, []string{
+		"ENV R8_PYTHON_VERSION=" + g.Config.Build.PythonVersion,
+	}...)
+
+	torchVersion, ok := g.Config.TorchVersion()
+	if ok {
+		lines = append(lines, []string{
+			"ENV R8_TORCH_VERSION=" + torchVersion,
+		}...)
+	}
+
+	buildTmpMount, err := g.buildTmpMount(tmpDir)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(lines, []string{
+		"RUN " + strings.Join([]string{
+			buildTmpMount,
+			g.monobaseUsercacheMount(),
+			APT_CACHE_MOUNT,
+			PIP_CACHE_MOUNT,
+		}, " ") + " /opt/r8/monobase/run.sh monobase.build " + skipCudaArg + " --mini --cache=" + MONOBASE_CACHE_PATH,
+	}...), nil
+}
+
+func (g *FastGenerator) copyWeights(lines []string, weights []Weight) ([]string, error) {
+	if len(weights) == 0 {
+		return lines, nil
+	}
+
+	for _, weight := range weights {
+		lines = append(lines, "COPY --link \""+weight.Path+"\" \""+filepath.Join(FUSE_RPC_WEIGHTS_PATH, weight.Digest)+"\"")
+	}
+
+	return lines, nil
+}
+
+func (g *FastGenerator) install(lines []string, weights []Weight, tmpDir string) ([]string, error) {
+	// Install apt packages
+	packages := g.Config.Build.SystemPackages
+	if len(packages) > 0 {
+		lines = append(lines, "RUN "+APT_CACHE_MOUNT+" apt-get update && apt-get install -qqy "+strings.Join(packages, " ")+" && rm -rf /var/lib/apt/lists/*")
+	}
+
+	// Install python packages
+	requirementsFile, err := g.pythonRequirements(tmpDir)
+	if err != nil {
+		return nil, err
+	}
+	buildTmpMount, err := g.buildTmpMount(tmpDir)
+	if err != nil {
+		return nil, err
+	}
+	if requirementsFile != "" {
+		lines = append(lines, "RUN "+strings.Join([]string{
+			buildTmpMount,
+			g.monobaseUsercacheMount(),
+			APT_CACHE_MOUNT,
+			PIP_CACHE_MOUNT,
+		}, " ")+" /opt/r8/monobase/run.sh monobase.build --requirements=/buildtmp/requirements.txt --mini --skip-cuda --cache="+MONOBASE_CACHE_PATH)
```

Comment:
We don't need `APT_CACHE_MOUNT` here right?

I decoupled this from mini-mono build to ensure it doesn't touch the existing layer.
So the command should be: `/opt/r8/monobase/run.sh monobase.user --requirements /src/requirements.txt`

We might want to mount `/srv/r8/monobase/uv/cache` with as a cache bind mount to work around this:
https://github.com/replicate/monobase/pull/63#issuecomment-2591478537

We might also need to exclude anything outside of `/root/.venv` to make sure the resulting tarball doesn't fail to write into read-only `/srv/r8/monobase` in K8S.

Reply:
Resolved in [0e019cb](https://github.com/replicate/cog/pull/2108/commits/0e019cb77300143a81a59328184501725c7eabe5)